src/model/ValveTanh.h

```diff
@@ -0,0 +1,197 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+/**
+ * @file ValveTanh.h
+ * @brief model::ValveTanh source file
+ */
+#ifndef SVZERODSOLVER_MODEL_VALVETANH_HPP_
+#define SVZERODSOLVER_MODEL_VALVETANH_HPP_
+
+#include <math.h>
+
+#include "Block.h"
+#include "SparseSystem.h"
+#include "debug.h"
+
+/**
+ * @brief Valve (tanh) block.
+ *
+ * Models the pressure drop across a diode-like valve, which is implemented as a
+ * non-linear hyperbolic-tangent resistor. See \cite pfaller2019importance.
+ *
+ * \f[
+ * \begin{circuitikz} \draw
+ * node[left] {$Q_{in}$} [-latex] (0,0) -- (0.8,0);
+ * \draw (1,0) node[anchor=south]{$P_{in}$}
+ * to [D, l=$R_v$, *-*] (3,0)
+ * node[anchor=south]{$P_{out}$};
+ * \end{circuitikz}
+ * \f]
+ *
+ * ### Governing equations
+ *
+ * \f[
+ * P_{in}-P_{out}-Q_{in}\left[R_{min} +
+ * (R_{max}-R_{min})\frac{1}{2}\left[1+tanh\{k(P_{out}-P{in})\}\right]\right]=0
+ * \f]
+ *
+ * \f[
+ * Q_{in}-Q_{out}=0
+ * \f]
+ *
+ * ### Local contributions
+ *
+ * \f[
+ * \mathbf{y}^{e}=\left[\begin{array}{llll}P_{in} & Q_{in} &
+ * P_{out} & Q_{out}\end{array}\right]^{T} \f]
+ *
+ * \f[
+ * \mathbf{E}^{e}=\left[\begin{array}{cccc}
+ * 0 & 0 & 0 & 0 \\
+ * 0 & 0 & 0 & 0
+ * \end{array}\right]
+ * \f]
+ *
+ * \f[
+ * \mathbf{F}^{e}=\left[\begin{array}{cccc}
+ * 1 & -(R_{max}+R_{min})/2.0 & -1 & 0 \\
+ * 0 &      1                 &  0 & -1
+ * \end{array}\right]
+ * \f]
+ *
+ * \f[
+ * \mathbf{c}^{e}=\left[\begin{array}{c}
+ * -\frac{1}{2}Q_{in}(R_{max}-R_{min})tanh\{k(P_{out}-P_{in})\} \\
+ * 0
+ * \end{array}\right]
+ * \f]
+ *
+ * \f[
+ * \left(\frac{\partial\mathbf{c}}{\partial\mathbf{y}}\right)^{e} =
+ * \left[\begin{array}{cccc}
+ * A & B & C & 0 \\
+ * 0 & 0 & 0 & 0 \end{array}\right] \f]
+ * where,
+ * \f[
+ * A = \frac{1}{2} k Q_{in}
+ * (R_{max}-R_{min})\left[1-tanh^2\{k(P_{out}-P_{in})\}\right] \\
+ * \f]
+ * \f[
+ * B = -\frac{1}{2}(R_{max}-R_{min})tanh\{k(P_{out}-P_{in})\} \\
+ * \f]
+ * \f[
+ * C = -\frac{1}{2} k Q_{in}
+ * (R_{max}-R_{min})\left[1-tanh^2\{k(P_{out}-P_{in})\}\right] \f]
+ *
+ * \f[
+ * \left(\frac{\partial\mathbf{c}}{\partial\dot{\mathbf{y}}}\right)^{e} =
+ * \left[\begin{array}{cccc}
+ * 0 & 0 & 0 & 0 \\
+ * 0 & 0 & 0 & 0
+ * \end{array}\right]
+ * \f]
+ *
+ * ### Parameters
+ *
+ * Parameter sequence for constructing this block
+ *
+ * * `0` Maximum (closed) valve resistance
+ * * `1` Minimum (open) valve resistance
+ * * `2` Steepness of sigmoid function
+ *
+ */
+class ValveTanh : public Block {
+ public:
+  /**
+   * @brief Local IDs of the parameters
+   *
+   */
+  enum ParamId {
```

Comment:
Yeah I think this would be great so that the names are consistent. Not sure how to do it either though. It might need something like the `BlockFactory` but a `ParamFactory`?